File: app/src/main/java/hcm/ditagis/com/cholon/qlsc/async/EditAsync.java
package hcm.ditagis.com.cholon.qlsc.async;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.graphics.Bitmap;
import android.os.AsyncTask;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;

import com.esri.arcgisruntime.concurrent.ListenableFuture;
import com.esri.arcgisruntime.data.ArcGISFeature;
import com.esri.arcgisruntime.data.Attachment;
import com.esri.arcgisruntime.data.CodedValue;
import com.esri.arcgisruntime.data.CodedValueDomain;
import com.esri.arcgisruntime.data.FeatureEditResult;
import com.esri.arcgisruntime.data.FeatureType;
import com.esri.arcgisruntime.data.Field;
import com.esri.arcgisruntime.data.ServiceFeatureTable;

import java.util.HashMap;
import java.util.List;
import java.util.concurrent.ExecutionException;

import hcm.ditagis.com.cholon.qlsc.adapter.FeatureViewMoreInfoAdapter;
import hcm.ditagis.com.cholon.qlsc.entities.DApplication;
import hcm.ditagis.com.cholon.qlsc.utities.Constant;


/**
 * Created by ThanLe on 4/16/2018.
 */

public class EditAsync extends AsyncTask<FeatureViewMoreInfoAdapter, ArcGISFeature, Void> {
    @SuppressLint("StaticFieldLeak")
    private ServiceFeatureTable mServiceFeatureTable;
    private ArcGISFeature mSelectedArcGISFeature;
    private byte[] mImage;
    private AsyncResponse mDelegate;
    private DApplication mApplication;
    private LinearLayout mLLayoutField;
    private HashMap<String, Object> mAttributes;
    private boolean mIsComplete;

    public interface AsyncResponse {
        void processFinish(ArcGISFeature feature);
    }

    public EditAsync(Activity activity, boolean isComplete,
                     ArcGISFeature selectedArcGISFeature, LinearLayout layout, byte[] image,
                     AsyncResponse delegate) {
        mLLayoutField = layout;
        mIsComplete = isComplete;
        mApplication = (DApplication) activity.getApplication();
        mServiceFeatureTable = (ServiceFeatureTable) selectedArcGISFeature.getFeatureTable();
        this.mDelegate = delegate;
        mSelectedArcGISFeature = selectedArcGISFeature;
        this.mImage = image;
    }

    @Override
    protected void onPreExecute() {
        super.onPreExecute();
        if (mImage == null) mAttributes = getAttributes();

    }

    private HashMap<String, Object> getAttributes() {
        HashMap<String, Object> attributes = new HashMap<>();
//        try {
        String currentAlias = "";
        for (int i = 0; i < mLLayoutField.getChildCount(); i++) {
            LinearLayout itemAddFeature = (LinearLayout) mLLayoutField.getChildAt(i);
            for (int j = 0; j < itemAddFeature.getChildCount(); j++) {
                LinearLayout typeInput_itemAddFeature = (LinearLayout) itemAddFeature.getChildAt(j);
                for (int k = 0; k < typeInput_itemAddFeature.getChildCount(); k++) {
                    View view = typeInput_itemAddFeature.getChildAt(k);
                    if (view.getVisibility() == View.VISIBLE) {
                        if (view instanceof EditText && !currentAlias.isEmpty()) {
                            for (Field field : mServiceFeatureTable.getFields()) {
                                if (field.getAlias().equals(currentAlias)) {
                                    if (field.getDomain() != null) {
                                        List<CodedValue> codedValues = ((CodedValueDomain) field.getDomain()).getCodedValues();

                                        Object valueDomain = getCodeDomain(codedValues, ((EditText) view).getText().toString());
                                        if (valueDomain != null)
                                            attributes.put(currentAlias, valueDomain.toString());
                                    } else {
                                        attributes.put(currentAlias, ((EditText) view).getText().toString());
                                    }
                                    break;
                                }
                            }
                        } else if (view instanceof Spinner && !currentAlias.isEmpty()) {
                            for (Field field : mServiceFeatureTable.getFields()) {
                                if (field.getAlias().equals(currentAlias)) {
                                    if (field.getDomain() != null) {
                                        List<CodedValue> codedValues = ((CodedValueDomain) field.getDomain()).getCodedValues();

                                        Object valueDomain = getCodeDomain(codedValues, ((Spinner) view).getSelectedItem().toString());
                                        if (valueDomain != null)
                                            attributes.put(currentAlias, valueDomain.toString());
                                    } else {
                                    }
                                    break;
                                }
                            }
                        } else if (view instanceof TextView) {
                            currentAlias = ((TextView) view).getText().toString();
                            attributes.put(currentAlias, null);
                        }
                    }
                }
            }
        }

//        } catch (Exception e) {
//            Log.e("Lỗi lấy attributes", e.toString());
//        }
        return attributes;
    }

    @Override
    protected Void doInBackground(FeatureViewMoreInfoAdapter... params) {
        if (mImage == null) {
            for (String alias : mAttributes.keySet()) {
                for (Field field : mServiceFeatureTable.getFields()) {
                    if (field.getAlias().equals(alias)) {
                        try {
                            String value = mAttributes.get(alias).toString().trim();
//                            if (value.isEmpty())
//                                continue;
                            switch (field.getFieldType()) {
                                case TEXT:
                                    mSelectedArcGISFeature.getAttributes().put(field.getName(), value);
                                    break;
                                case DOUBLE:
                                    mSelectedArcGISFeature.getAttributes().put(field.getName(), Double.parseDouble(value));
                                    break;
                                case FLOAT:
                                    mSelectedArcGISFeature.getAttributes().put(field.getName(), Float.parseFloat(value));
                                    break;
                                case INTEGER:
                                    mSelectedArcGISFeature.getAttributes().put(field.getName(), Integer.parseInt(value));
                                    break;
                                case SHORT:
                                    mSelectedArcGISFeature.getAttributes().put(field.getName(), Short.parseShort(value));
                                    break;
                            }

                        } catch (Exception e) {
                            mSelectedArcGISFeature.getAttributes().put(field.getName(), null);
                            Log.e("Lỗi thêm điểm", e.toString());
                        }
                        break;
                    }
                }
            }
        }
        if (mIsComplete)
            mSelectedArcGISFeature.getAttributes().put(Constant.FieldSuCo.TRANG_THAI, Constant.TrangThaiSuCo.HOAN_THANH);
        mServiceFeatureTable.loadAsync();
        mServiceFeatureTable.addDoneLoadingListener(() -> {
            // update feature in the feature table
            mServiceFeatureTable.updateFeatureAsync(mSelectedArcGISFeature).addDoneListener(() ->
                    mServiceFeatureTable.applyEditsAsync().addDoneListener(() -> {
                        if (mImage != null) {
                            if (mSelectedArcGISFeature.canEditAttachments())
                                addAttachment();
                            else
                                applyEdit();
                        } else {
                            applyEdit();

                        }
                    }));
        });
        return null;
    }

    private void addAttachment() {
        final String attachmentName = "attachment_" + System.currentTimeMillis() + ".png";
        final ListenableFuture<Attachment> addResult = mSelectedArcGISFeature.addAttachmentAsync(mImage, Bitmap.CompressFormat.PNG.toString(), attachmentName);
        addResult.addDoneListener(() -> {
            try {
                Attachment attachment = addResult.get();
                if (attachment.getSize() > 0) {
                    final ListenableFuture<Void> tableResult = mServiceFeatureTable.updateFeatureAsync(mSelectedArcGISFeature);
                    tableResult.addDoneListener(() -> applyEdit());
                }
            } catch (Exception ignored) {
                publishProgress();
            }
        });
    }


    private void applyEdit() {

        final ListenableFuture<List<FeatureEditResult>> updatedServerResult = mServiceFeatureTable.applyEditsAsync();
        updatedServerResult.addDoneListener(() -> {
            try {
                updatedServerResult.get();
                publishProgress(mSelectedArcGISFeature);
            } catch (InterruptedException | ExecutionException e) {
                e.printStackTrace();
                publishProgress();
            }

        });

    }

    private Object getIdFeatureTypes(List<FeatureType> featureTypes, String value) {
        Object code = null;
        for (FeatureType featureType : featureTypes) {
            if (featureType.getName().equals(value)) {
                code = featureType.getId();
                break;
            }
        }
        return code;
    }

    private Object getCodeDomain(List<CodedValue> codedValues, String value) {
        Object code = null;
        for (CodedValue codedValue : codedValues) {
            if (codedValue.getName().equals(value)) {
                code = codedValue.getCode();
                break;
            }
        }
        return code;
    }

    @Override
    protected void onProgressUpdate(ArcGISFeature... values) {
        super.onProgressUpdate(values);
        if (values != null && values.length > 0)
            this.mDelegate.processFinish(values[0]);
        else this.mDelegate.processFinish(null);
    }

    @Override
    protected void onPostExecute(Void result) {
        super.onPostExecute(result);

    }

}

File: app/src/main/java/hcm/ditagis/com/cholon/qlsc/async/EditAsync.kt
package hcm.ditagis.com.cholon.qlsc.async

import android.annotation.SuppressLint
import android.app.Activity
import android.os.AsyncTask
import android.util.Log
import android.view.View
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.Spinner
import android.widget.TextView
import com.esri.arcgisruntime.data.*
import hcm.ditagis.com.cholon.qlsc.entities.DApplication
import hcm.ditagis.com.cholon.qlsc.utities.Constant
import java.util.*
import java.util.concurrent.ExecutionException


/**
 * Created by ThanLe on 4/16/2018.
 */

class EditAsync(activity: Activity, private val mIsComplete: Boolean,
                private val mSelectedArcGISFeature: ArcGISFeature, private val mLLayoutField: LinearLayout?, private val mImage: ByteArray?,
                private val mDelegate: AsyncResponse) : AsyncTask<Void, ArcGISFeature, Void>() {
    @SuppressLint("StaticFieldLeak")
    private val mServiceFeatureTable: ServiceFeatureTable
    private val mApplication: DApplication
    private var mAttributes: HashMap<String, Any?>? = null

    private//        try {
    //        } catch (Exception e) {
    //            Log.e("Lỗi lấy attributes", e.toString());
    //        }
    val attributes: HashMap<String, Any?>
        get() {
            val attributes = HashMap<String, Any?>()
            var currentAlias = ""
            if (mLLayoutField != null)
                for (i in 0 until mLLayoutField.childCount) {
                    val itemAddFeature = mLLayoutField.getChildAt(i) as LinearLayout
                    for (j in 0 until itemAddFeature.childCount) {
                        val typeInput_itemAddFeature = itemAddFeature.getChildAt(j) as LinearLayout
                        for (k in 0 until typeInput_itemAddFeature.childCount) {
                            val view = typeInput_itemAddFeature.getChildAt(k)
                            if (view.visibility == View.VISIBLE) {
                                if (view is EditText && !currentAlias.isEmpty()) {
                                    for (field in mServiceFeatureTable.fields) {
                                        if (field.alias == currentAlias) {
                                            if (field.domain != null) {
                                                val codedValues = (field.domain as CodedValueDomain).codedValues

                                                val valueDomain = getCodeDomain(codedValues, view.text.toString())
                                                if (valueDomain != null)
                                                    attributes[currentAlias] = valueDomain.toString()
                                            } else {
                                                attributes[currentAlias] = view.text.toString()
                                            }
                                            break
                                        }
                                    }
                                } else if (view is Spinner && !currentAlias.isEmpty()) {
                                    for (field in mServiceFeatureTable.fields) {
                                        if (field.alias == currentAlias) {
                                            if (field.domain != null) {
                                                val codedValues = (field.domain as CodedValueDomain).codedValues

                                                val valueDomain = getCodeDomain(codedValues, view.selectedItem.toString())
                                                if (valueDomain != null)
                                                    attributes[currentAlias] = valueDomain.toString()
                                            } else {
                                            }
                                            break
                                        }
                                    }
                                } else if (view is TextView) {
                                    currentAlias = view.text.toString()
                                    attributes[currentAlias] = null
                                }
                            }
                        }
                    }
                }
            return attributes
        }

    interface AsyncResponse {
        fun processFinish(feature: ArcGISFeature?)
    }

    init {
        mApplication = activity.application as DApplication
        mServiceFeatureTable = mSelectedArcGISFeature.featureTable as ServiceFeatureTable
    }

    override fun onPreExecute() {
        super.onPreExecute()
        if (mImage == null) mAttributes = attributes

    }

    override fun doInBackground(vararg params: Void): Void? {
        if (mImage == null) {
            for (alias in mAttributes!!.keys) {
                for (field in mServiceFeatureTable.fields) {
                    if (field.alias == alias) {
                        try {
                            val value = mAttributes!![alias].toString().trim { it <= ' ' }
                            //                            if (value.isEmpty())
                            //                                continue;
                            when (field.fieldType) {
                                Field.Type.TEXT -> mSelectedArcGISFeature.attributes[field.name] = value
                                Field.Type.DOUBLE -> mSelectedArcGISFeature.attributes[field.name] = java.lang.Double.parseDouble(value)
                                Field.Type.FLOAT -> mSelectedArcGISFeature.attributes[field.name] = java.lang.Float.parseFloat(value)
                                Field.Type.INTEGER -> mSelectedArcGISFeature.attributes[field.name] = Integer.parseInt(value)
                                Field.Type.SHORT -> mSelectedArcGISFeature.attributes[field.name] = java.lang.Short.parseShort(value)
                            }

                        } catch (e: Exception) {
                            mSelectedArcGISFeature.attributes[field.name] = null
                            Log.e("Lỗi thêm điểm", e.toString())
                        }

                        break
                    }
                }
            }
        }
        if (mIsComplete)
            mSelectedArcGISFeature.attributes[Constant.FieldSuCo.TRANG_THAI] = Constant.TrangThaiSuCo.HOAN_THANH
        mServiceFeatureTable.loadAsync()
        mServiceFeatureTable.addDoneLoadingListener {
            // update feature in the feature table
            mServiceFeatureTable.updateFeatureAsync(mSelectedArcGISFeature).addDoneListener {
                mServiceFeatureTable.applyEditsAsync().addDoneListener {
                    if (mImage != null) {
                        if (mSelectedArcGISFeature.canEditAttachments())
                            addAttachment()
                        else
                            applyEdit()
                    } else {
                        applyEdit()

                    }
                }
            }
        }
        return null
    }

    private fun addAttachment() {
        val attachmentName = String.format(Constant.AttachmentName.UPDATE,
                mApplication.userDangNhap!!.userName, System.currentTimeMillis())
        val addResult = mSelectedArcGISFeature.addAttachmentAsync(mImage, Constant.FileType.PNG, attachmentName)
        addResult.addDoneListener {
            try {
                val attachment = addResult.get()
                if (attachment.size > 0) {
                    val tableResult = mServiceFeatureTable.updateFeatureAsync(mSelectedArcGISFeature)
                    tableResult.addDoneListener { applyEdit() }
                }
            } catch (ignored: Exception) {
                publishProgress()
            }
        }
    }


    private fun applyEdit() {

        val updatedServerResult = mServiceFeatureTable.applyEditsAsync()
        updatedServerResult.addDoneListener {
            try {
                updatedServerResult.get()
                publishProgress(mSelectedArcGISFeature)
            } catch (e: InterruptedException) {
                e.printStackTrace()
                publishProgress()
            } catch (e: ExecutionException) {
                e.printStackTrace()
                publishProgress()
            }


        }

    }

    private fun getIdFeatureTypes(featureTypes: List<FeatureType>, value: String): Any? {
        var code: Any? = null
        for (featureType in featureTypes) {
            if (featureType.name == value) {
                code = featureType.id
                break
            }
        }
        return code
    }

    private fun getCodeDomain(codedValues: List<CodedValue>, value: String): Any? {
        var code: Any? = null
        for (codedValue in codedValues) {
            if (codedValue.name == value) {
                code = codedValue.code
                break
            }
        }
        return code
    }

    override fun onProgressUpdate(vararg values: ArcGISFeature) {
        super.onProgressUpdate(*values)
        if (values != null && values.size > 0)
            this.mDelegate.processFinish(values[0])
        else
            this.mDelegate.processFinish(null)
    }

}

